infrastructure/storage/src/lmdb_store/store.rs

```diff
@@ -0,0 +1,586 @@
+//! An ergonomic, multithreaded API for an LMDB datastore
+
+use crate::lmdb_store::error::LMDBError;
+use lmdb_zero::{
+    db,
+    error::{self, LmdbResultExt},
+    open,
+    put,
+    traits::AsLmdbBytes,
+    ConstAccessor,
+    Cursor,
+    CursorIter,
+    Database,
+    DatabaseOptions,
+    EnvBuilder,
+    Environment,
+    Ignore,
+    MaybeOwned,
+    ReadTransaction,
+    Stat,
+    WriteAccessor,
+    WriteTransaction,
+};
+use log::*;
+use std::{cmp::max, collections::HashMap, path::Path, sync::Arc};
+
+const LOG_TARGET: &str = "lmdb";
+
+/// An atomic pointer to an LMDB database instance
+type DatabaseRef = Arc<Database<'static>>;
+
+/// A builder for [LMDBStore](struct.lmdbstore.html)
+/// ## Example
+///
+/// Create a new LMDB database of 500MB in the `db` directory with two named databases: "db1" and "db2"
+///
+/// ```
+/// # use tari_storage::lmdb_store::LMDBBuilder;
+/// # use lmdb_zero::db;
+/// let mut store = LMDBBuilder::new()
+///     .set_path("/tmp/")
+///     .set_mapsize(500)
+///     .set_max_dbs(10)
+///     .add_database("db1", db::CREATE)
+///     .add_database("db2", db::CREATE)
+///     .build()
+///     .unwrap();
+/// ```
+pub struct LMDBBuilder {
+    path: String,
+    db_size_mb: usize,
+    max_dbs: usize,
+    db_names: HashMap<String, db::Flags>,
+}
+
+impl LMDBBuilder {
+    /// Create a new LMDBStore builder. Set up the database by calling `set_nnnn` and then create the database
+    /// with `build()`. The default values for the database parameters are:
+    ///
+    /// | Parameter | Default |
+    /// |:----------|---------|
+    /// | path      | ./store/|
+    /// | size      | 64 MB   |
+    /// | named DBs | none    |
+    pub fn new() -> LMDBBuilder {
+        LMDBBuilder {
+            path: "./store/".into(),
+            db_size_mb: 64,
+            db_names: HashMap::new(),
+            max_dbs: 8,
+        }
+    }
+
+    /// Set the directory where the LMDB database exists, or must be created.
+    /// Note: The directory must exist already; it is not created for you. If it does not exist, `build()` will
+    /// return `LMDBError::InvalidPath`.
+    pub fn set_path(mut self, path: &str) -> LMDBBuilder {
+        self.path = path.into();
+        if !self.path.ends_with("/") {
+            self.path += "/";
+        }
+        self
+    }
+
+    /// Sets the size of the database, in MB.
+    /// The actual memory will only be allocated when #build() is called
+    pub fn set_mapsize(mut self, size: usize) -> LMDBBuilder {
```

Comment:
Would this not be more clear to just make this: `set_db_size`

Reply:
Sure. Can change that.